### ISSUE TITLE
Sequence doesn't match the reference implementation with the same seed

### ISSUE THREAD

**Issue by alisey** [thomcc/pcg-random#2]
Calling [pcg_random()](https://github.com/imneme/pcg-c-basic/blob/master/pcg_basic.c#L69) from the [official minimal C implementation](https://www.pcg-random.org/using-pcg-c-basic.html) three times in a row produces: `0x152ca78d`, `0x027c6003`, `0xcb07bbf3`. But calling the JS version with the same seed produces a different sequence.

Here's a [sandbox](https://codesandbox.io/s/pcg-random-comparison-rgs7x?file=/src/index.js) comparing the output of `pcg-random` to the expected output. I've also added a naive implementation based on `BigInt`, which matches the expected output. I didn't investigate deeply, but one issue is that `mul64_` isn't working correctly:
```js
const out = [12, 34];
mul64_(out, 1, 0, 1, 0);
console.log(out); // [0, 0]
```

**Comment by thomcc** [comment]
Yeah, hm. I'll take a look in a few days, sorry for the delay. That's concerning, for sure.

**Comment by thomcc** [comment]
This will be fixed in #3 (which I'll finish tomorrow probably). Very sorry, it's honestly kind of an embarrassing bug to have[^1]. Thanks a bunch for reporting it.

[^1]: Especially given that I've shipped code that used this...

I even vaguely remember thinking that I wish pcg-random included test sequences, and meaning to make them myself... but it didn't (and still doesn't, afaict), and I never got around to it.

I generated a few of these manually in #3, and don't *really* feel like doing the work to write a more robust test framework (such as one that compiles some C code and tests that the output is equivalent for a bunch of random inputs), but maybe I'll feel differently tomorrow.

I think I'm going to consider this a breaking[^2] change, and use the opportunity to fix some other things[^3]. It also probably could use a bit more docs (which that PR starts on).

[^2]: Since a hard-coded seed will no longer produce the same sequence.

[^3]: Like the use of big-endian for arguments, and in the state array (not sure why I did that), and such. That said, I don't intend on using any JS features or w/e that the current version doesn't use. (IDK if anybody cares about runtimes/browsers that old though. Certainly, I'd recommend against caring, but it's easy to support).

---

Since I guess you're a user, or a potential one... While you're here, are RNG utilities like (shuffle functions, better "random int in range" support, and etc) compelling features? I have a handful of these since I used a (err.... copypasted version of) this in some games a while ago. It would be easy for me to include, but I'm unsure if it's considered useful.

Same question for the PCG "tricks", like reversing and skipping ahead (I'm assuming you're at least a little familiar with the PCG featureset, since you linked to the C code). This would be a bit more work to support, but not too bad.

(I'd rather not do these as micropackages though).

Obviously this stuff would be better tested than the existing code. (This code is like 7 years old, and since then I've... developed significantly better habits around testing)

**Comment by thomcc** [comment]
I've published the fixed version as [v2.0.0](https://github.com/thomcc/pcg-random/releases/tag/v2.0.0). Still interested in hearing your answers to the questions above, if you're interested in answering them.

If not, thanks for the report, sorry for taking so long to get to it (holidays), and for having the bug in the first place (it really isn't the kind of bug that you should have).

**Comment by alisey** [comment]
Thank you!

> Are RNG utilities compelling features?

In general, I like the idea of decoupling random number engines from utilities that use them, like in C++ or in [D3](https://github.com/d3/d3-random):

```js
const random = randomNormal.source(myRng)(0, 1);
random(); // -0.6253955998897069
```

At first glance it seems unergonomic to specify an rng every time you call `shuffle()`, since you don't normally care about the engine used.

But if you think of a scale of utilities going from simple to complex: random integer, normal distribution, simplex noise, at some point it becomes unreasonable to tie them to a specific engine. And as a library user you don't want to end up in a place where your `shuffle()` is using PCG, but you `reservoirSample()` is using ChaCha.

In [my own implementation](https://github.com/alisey/pcg32) of PCG I hesitated whether to provide a function for generating random floats, but decided to add it because it is at the same time very useful as a replacement for `Math.random()`, and tricky to implement despite being a one-liner.

> Reversing and skipping ahead

This could be useful for long-running simulations where you want to be able to continue after a crash or power outage, but JavaScript is not the most natural choice for those.

For my use (little toys and generative art), I never cared for those features.
